[FIX] hr_holidays: correct computation of new record

Abdessamad-z opened the pull request:
To reproduce the bug, follow these steps:

1. Create a new time off type.
2. Create a new accrual plan.
3. Set the "Accrued Gain Time" field to “At the start of the accrual period.”
4. Create a new milestone with a cap.
5. Create a new allocation set its type to accrual then link it to our accrual plan.
6. Set the start date in the past so that the accrued days reach the cap.
7. Create a new leave in the past related to the allocation.
8. Check your balance for a future date in the Time Off app.

Normally, with accrual allocation, we should see the allocation increase. The problem here is that in the future, the
allocation decreases.

This issue arises due to the way the `leaves_taken` field is computed.

https://github.com/odoo/odoo/blob/08a4b9d78addeefde24b189f0066540e0dcee4c4/addons/hr_holidays/models/hr_leave_allocation.py#L215-L220

The value is retrieved from a default dictionary, which is processed based on existing records in the database. When we
attempt to fetch the value using a new instance that only exists in memory, the key returns the default value of 0.

Additionally, if we modify step 7 to set the leave in the future, the accrual value caps out at a level below the
intended maximum. In some cases, if the data-fetching function is called with the same parameters consecutively, we
receive different outputs.

These issues arise because the new record persists as a cache. When we retrieve a new record with the same origin as a
previous call, we receive the previously processed data. This also leads to a problem with recursion in the function
`_get_future_leaves_on`. This commit addresses and resolves these issues.

opw-4040882

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/174354.png)](https://mergebot.odoo.com/odoo/odoo/pull/174354)

abla001 reviewed (approved):
LGTM!

Bertrand2 reviewed (commented):
LGTM! :)
IMO It just misses a check on the number of days of the allocation upon creation, to ensure that before the leave, nothing could be added

Abdessamad-z commented:
Thanks for the review! I added the suggested change.

Bertrand2 commented on addons/hr_holidays/tests/test_accrual_allocations.py:
```diff
@@ -1846,3 +1846,55 @@ def test_check_lastcall_change_regular_to_accrual(self):
+            record_data = new_allocation._convert_to_write(new_allocation._cache)
+            allocation = self.env['hr.leave.allocation'].create(record_data)
+            allocation.action_validate()
+            accrual_bonus = allocation._get_future_leaves_on(datetime.date(2024, 3, 2))
```
I think it should check the same date as below, march 3rd

Bertrand2 reviewed (commented):
Thanks for the change! Here's one more comment :)
Otherwise I have a question regarding your commit message: you mention in the steps to "Update the fields “lastcall” and “nextcall” ...". However, those fields aren't accessible functionally, so I'm not sure of what you mean by that

Abdessamad-z commented:
Effectively you can't access them directly. I updated them using python shell. You can always change the local device date (I added that to the commit message), but it's inconvenient.

Bertrand2 reviewed (changes requested):
Thanks for the fix! 😀 I just noticed two things that need to be addressed for it to be perfect:
- The test doesn't really require to run the accrual plan. The whole idea behind the future leaves is that it simulates it without actually running the cron; so IMO the test shouldn't move between multiple dates, it could be reproduced like this:
  1. Have an accrual plan with a cap (e.g. 1 day granted every day, cap at 10)
  1. create an allocation starting in the past so that the cap is reached (e.g. 1 month ago)
  1. create a leave (e.g. yesterday)
  1. see the future balance on a date after the leave that should grant new days (e.g. tomorrow) 
- The fix is actually incomplete: for past leaves this works perfectly, but I noticed that it doesn't work for future leaves (to test that, you can create the leave in step iii tomorrow instead of yesterday).
The way accrual allocations work is that future leaves are not taken into account in the current balance, so just giving as `leaves_taken` the current `leaves_taken` balance will ignore those. I guess you could add the duration of the leaves that start between today and the date at which we want to see the balance, but I haven't checked what's the current behavior for leaves and the balance if the accrual is based on worked time

Abdessamad-z commented:
@Bertrand2 Hello thanks for the propositions! What do you think about these changes?

Bertrand2 commented on addons/hr_holidays/models/hr_leave_allocation.py:
```diff
@@ -489,11 +497,18 @@ def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
             # get current level and normal period boundaries, then set nextcall, adjusted for level transition and carryover
             # add days, trimmed if there is a maximum_leave
             while allocation.nextcall <= date_to:
+                is_leave = False
```
Why not grouping everything using this variable (declaration, assignation and condition) to make it clearer? Also, I'm not sure that the variable name is the best; it's linked to the allocation, and it's relative to having future leaves :thinking:

Bertrand2 commented on addons/hr_holidays/models/hr_leave_allocation.py:
```diff
@@ -531,9 +546,13 @@ def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
             # once, preventing double allocation.
             if allocation.accrual_plan_id.accrued_gain_time == 'start':
                 # check that we are at the start of a period, not on a carry-over or level transition date
+                is_leave = False
```
unused?

Abdessamad-z replied:
It is used below

Bertrand2 commented on addons/hr_holidays/models/hr_leave_allocation.py:
```diff
@@ -569,6 +588,7 @@ def _get_future_leaves_on(self, accrual_date):
                 and (not self.nextcall or self.nextcall <= accrual_date)):
             return 0
 
+        self._invalidate_cache()
```
Why is it needed?

Abdessamad-z replied:
As i explained in the commit message, if we create a new record it first browse the cache looking if the NewId is already existant. If it is, we get the value from cache, meaning if we did some treatment on  them it persist for the second call. So calling the function 2 consecutive time with same parameter can have different output.

Bertrand2 commented on addons/hr_holidays/tests/test_accrual_allocations.py:
```diff
@@ -1846,3 +1846,73 @@ def test_check_lastcall_change_regular_to_accrual(self):
+                })],
+            })
+            # Simulate creating an allocation from frontend interface
+            new_allocation = self.env['hr.leave.allocation'].new({
+                'name': 'Accrual allocation for employee',
+                'accrual_plan_id': accrual_plan.id,
+                'employee_id': self.employee_emp.id,
+                'holiday_status_id': leave_type.id,
+                'date_from': '2024-02-01',
+                'allocation_type': 'accrual',
+                'holiday_type': 'employee',
+            })
+            new_allocation._onchange_date_from()
+            record_data = new_allocation._convert_to_write(new_allocation._cache)
+            allocation = self.env['hr.leave.allocation'].create(record_data)
+            allocation.action_validate()
```
use `Form` instead, like in `test_accrual_period_start_past_start_date`

Bertrand2 commented on addons/hr_holidays/tests/test_accrual_allocations.py:
```diff
@@ -1846,3 +1846,73 @@ def test_check_lastcall_change_regular_to_accrual(self):
+            leave.action_validate()
+            self.assertEqual(get_reamining_leaves(2024, 3, 9), 10, "Leaves following the date selected should not be counted")
+            self.assertEqual(get_reamining_leaves(2024, 3, 10), 5, "5 days should be deduced from the allocation")
+            self.assertEqual(get_reamining_leaves(2024, 3, 15), 5, "No bonus should be added during a leave")
```
If the period at which the days are accrued is "start", we shouldn't take into account anything upon accruing days. The whole point of that is to not know what is coming :) So in this case, some days should have been accrued again after the 10th, because the balance has been impacted by the leave having started

Abdessamad-z commented:
hello @Bertrand2 , what do you think about the current changes

Bertrand2 reviewed (approved):
LGTM! It seems to work as well.
Thanks!

Bertrand2 commented:
robodoo r+
